src/engine/controls/cuecontrol.cpp

```diff
@@ -1589,69 +1588,116 @@ CueControl::TrackAt CueControl::getTrackAt() const {
+        if (nextBeat != -1.0 && nextBeat <= total) {
+            return nextBeat;
+        }
+        return currentPos;
+    }
+    if (mode == QuantizeMode::NextBeat) {
+        // use current position if we are already exactly on the grid,
+        // otherwise quantize to next beat, fall back to previous beat.
+        if (prevBeat == currentPos) {
+            return currentPos;
+        }
+        if (nextBeat != -1.0 && nextBeat <= total) {
+            return nextBeat;
+        }
+        if (prevBeat != -1.0) {
+            return prevBeat;
+        }
+        return currentPos;
     }
+    DEBUG_ASSERT(!"PROGRAMMING ERROR: Invalid quantize mode");
+    return currentPos;
 }
 
-double CueControl::quantizeCuePoint(double position, Cue::CueSource source, QuantizeMode mode) {
+double CueControl::quantizeCuePoint(double cuePos, QuantizeMode mode) {
+    SampleOfTrack sampleOfTrack = getSampleOfTrack();
+    const double total = sampleOfTrack.total;
+
+    VERIFY_OR_DEBUG_ASSERT(cuePos <= total) {
+        cuePos = total;
+    }
+
     // Don't quantize unset cues, manual cues or when quantization is disabled.
-    if (position == -1.0 || source == Cue::MANUAL || !m_pQuantizeEnabled->toBool()) {
-        return position;
+    if (cuePos == Cue::kNoPosition || !m_pQuantizeEnabled->toBool()) {
+        return cuePos;
     }
 
     BeatsPointer pBeats = m_pLoadedTrack->getBeats();
     if (!pBeats) {
-        return position;
+        return cuePos;
     }
 
     if (mode == QuantizeMode::ClosestBeat) {
-        double closestBeat = pBeats->findClosestBeat(position);
-        return closestBeat != -1.0 ? closestBeat : position;
+        double closestBeat = pBeats->findClosestBeat(cuePos);
+        if (closestBeat != -1.0 && closestBeat <= total) {
+            return cuePos;
+        }
+        return cuePos;
     }
 
     double prevBeat, nextBeat;
-    pBeats->findPrevNextBeats(position, &prevBeat, &nextBeat);
+    pBeats->findPrevNextBeats(cuePos, &prevBeat, &nextBeat);
 
     if (mode == QuantizeMode::PreviousBeat) {
         // Quantize to previous beat, fall back to next beat.
-        return prevBeat != -1.0 ? prevBeat : (nextBeat != -1.0 ? nextBeat : position);
-    } else if (mode == QuantizeMode::NextBeat) {
-        // Quantize to next beat, fall back to previous beat.
-        return nextBeat != -1.0 ? nextBeat : (prevBeat != -1.0 ? prevBeat : position);
-    } else {
-        qWarning() << "PROGRAMMING ERROR: Invalid quantize mode" << static_cast<int>(mode);
-        return -1.0;
+        if (prevBeat != -1.0) {
+            return prevBeat;
+        }
+        if (nextBeat != -1.0 && nextBeat <= total) {
+            return nextBeat;
+        }
+        return cuePos;
     }
+    if (mode == QuantizeMode::NextBeat) {
+        // use current cuePos if we are already exactly on the grid,
+        // otherwise quantize to next beat, fall back to previous beat.
+        if (prevBeat == cuePos) {
+            return position;
```

Comment:
to fix build errors:
```
--- a/src/engine/controls/cuecontrol.cpp
+++ b/src/engine/controls/cuecontrol.cpp
@@ -1686,7 +1686,7 @@ double CueControl::quantizeCuePoint(double cuePos, QuantizeMode mode) {
         // use current cuePos if we are already exactly on the grid,
         // otherwise quantize to next beat, fall back to previous beat.
         if (prevBeat == cuePos) {
-            return position;
+            return cuePos;
         }
         if (nextBeat != -1.0 && nextBeat <= total) {

```